File: crates/harness-tools/src/echo.rs
use async_trait::async_trait;
use harness_core::{BoxError, Tool, ToolCategory};
use serde_json::{json, Value};

/// Trivial echo tool — returns its `text` argument verbatim. Useful for
/// smoke-testing the tool loop end to end.
pub struct EchoTool;

#[async_trait]
impl Tool for EchoTool {
    fn name(&self) -> &str {
        "echo"
    }

    fn description(&self) -> &str {
        "Echo the `text` argument back verbatim. Useful for testing the tool loop."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "text": { "type": "string", "description": "Text to echo." }
            },
            "required": ["text"]
        })
    }

    fn cacheable(&self) -> bool {
        true
    }

    fn category(&self) -> ToolCategory {
        ToolCategory::Read
    }

    async fn invoke(&self, args: Value) -> Result<String, BoxError> {
        let text = args
            .get("text")
            .and_then(Value::as_str)
            .ok_or_else(|| -> BoxError { "missing `text` argument".into() })?;
        Ok(text.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn echoes_text() {
        let out = EchoTool.invoke(json!({ "text": "hi" })).await.unwrap();
        assert_eq!(out, "hi");
    }

    #[tokio::test]
    async fn rejects_missing_text() {
        let err = EchoTool.invoke(json!({})).await.unwrap_err();
        assert!(err.to_string().contains("text"));
    }
}

File: packages/tools/src/echo.ts
// `echo` — trivial always-on tool. Ported from harness-tools/src/echo.rs.
//
// Returns its `text` argument back verbatim. Useful for smoke-testing the
// tool loop end to end.
import type { JsonValue } from "@jarvis/core";
import type { Tool, ToolCategory } from "@jarvis/core";

export class EchoTool implements Tool {
  readonly name = "echo";
  readonly description = "Echo the `text` argument back verbatim. Useful for testing the tool loop.";
  readonly cacheable = true;
  readonly category: ToolCategory = "read";

  readonly parameters: JsonValue = {
    type: "object",
    properties: {
      text: { type: "string", description: "Text to echo." },
    },
    required: ["text"],
  };

  async invoke(args: JsonValue): Promise<string> {
    const text =
      args !== null && typeof args === "object" && !Array.isArray(args)
        ? args["text"]
        : undefined;
    if (typeof text !== "string") {
      throw new Error("missing `text` argument");
    }
    return text;
  }
}
